library/core/src/slice/iter/macros.rs

```diff
@@ -1,36 +1,60 @@
 //! Macros used by iterators of slice.
 
-/// Convenience & performance macro for consuming the `end_or_len` field, by
+/// Convenience macro for updating the `end_addr_or_len` field for non-ZSTs.
+macro_rules! set_end {
+    ($this:ident . end = $new_end:expr) => {{
+        $this.end_addr_or_len = addr_usize($new_end);
+    }};
+}
+
+/// Convenience & performance macro for consuming the `end_addr_or_len` field, by
 /// giving a `(&mut) usize` or `(&mut) NonNull<T>` depending whether `T` is
 /// or is not a ZST respectively.
 ///
-/// Internally, this reads the `end` through a pointer-to-`NonNull` so that
-/// it'll get the appropriate non-null metadata in the backend without needing
-/// to call `assume` manually.
+/// When giving a `NonNull<T>` for the end, it creates it by offsetting from the
+/// `ptr` so that the backend knows that both pointers have the same provenance.
 macro_rules! if_zst {
     (mut $this:ident, $len:ident => $zst_body:expr, $end:ident => $other_body:expr,) => {{
         #![allow(unused_unsafe)] // we're sometimes used within an unsafe block
 
         if T::IS_ZST {
-            // SAFETY: for ZSTs, the pointer is storing a provenance-free length,
-            // so consuming and updating it as a `usize` is fine.
-            let $len = unsafe { &mut *ptr::addr_of_mut!($this.end_or_len).cast::<usize>() };
+            let $len = &mut $this.end_addr_or_len;
             $zst_body
         } else {
-            // SAFETY: for non-ZSTs, the type invariant ensures it cannot be null
-            let $end = unsafe { &mut *ptr::addr_of_mut!($this.end_or_len).cast::<NonNull<T>>() };
+            // SAFETY: By type invariant `end >= ptr`, and thus the subtraction
+            // cannot overflow, and the iter represents a single allocated
+            // object so the `add` will also be in-range.
+            let $end = unsafe {
+                let ptr_addr = addr_usize($this.ptr);
+                // Need to load as `NonZero` to get `!range` metadata
+                let end_addr: NonZero<usize> = *ptr::addr_of!($this.end_addr_or_len).cast();
+                // Not using `with_addr` because we have ordering information that
+                // we can take advantage of here that `with_addr` cannot.
+                let byte_diff = intrinsics::unchecked_sub(end_addr.get(), ptr_addr);
+                $this.ptr.byte_add(byte_diff)
+            };
```

Comment:
That's unfortunate. I think this will make #120682 impossible since there my approach for forward iteration is to only operate off `end` and never read `ptr`, only write to it. That makes the writes easy to hoist. If re-acquiring the provenance requires reading from `ptr` in each `next` that no longer works.

It'd be great if instead we could somehow tell LLVM that two pointers have the same provenance.

Reply:
Could we keep it a pointer so it can be used as such when it's useful and do ptr-addr-arithmetic-ptr dances otherwise? or would that lose the intended benefits?

Reply:
My goal with that PR is to make unchecked indexing less hazardous and also applicable to `vec::IntoIter` with `T: Drop` which would mean `zip(vec::IntoIter, slice::Iter)` would now get the unchecked indexing too in those cases which it didn't previously. So there would also be perf benefits from that approach.

Also, have you checked how this affects looping over `next_back` or `rev`? It seems that it would have to recalculate the length / pointer provenance on each iteration. In the end it might optimize away, but the IR probably becomes more gnarly.

> I think the problem with "when it's useful" is that it has to mean potentially-two-provenances

For those methods. But the other methods could treat the field as usize, no?